cataclysmbnteam/Cataclysm-BN
# feat(balance): uncanny dodge can now ninja roll to dodge bullets

## Purpose of change

![image](https://github.com/cataclysmbnteam/Cataclysm-BN/assets/54838975/db49a9e4-4a1c-42fd-85e9-d0d6ad99f6b4)

https://github.com/cataclysmbnteam/Cataclysm-BN/assets/54838975/8e2ba940-f7d1-42ec-b3ab-8351c179e878

it's #4125 but with ninja moving

## Describe the solution

on top of #4125, it adds...

- `src/cata_algo.h`: a very [convoluted template mess to use `std::variant` sanely](https://en.cppreference.com/w/cpp/utility/variant/visit)!

- `src/character_functions.h`: describe uncanny dodge simulation result in `std::variant`!
- `src/character_functions.cpp`:
	- separate uncanny dodge simulation and execution into `uncanny_dodge_result` and `try_uncanny_dodge`!
	- creates smoke when dodging to block sight and reduce adjacent tiles to dodge when `thruster mode` is active

## Testing

1. edit bionics, give max power and uncanny dodge CBM
2. spawn turret.
3. set dodge to 0, check that you dodge the bullets around 50% of the time
4. set dodge to 10, check that you dodge the bullets 100% of the time
5. set dodge to 0, enable `Uncanny Dodge Thrusters Mode`, check that you dodge the bullets 100% of the time with leaving smoke behind (unless you're not surrounded by unsafe terrain)
6. set dodge to 10, enable `Uncanny Dodge Thrusters Mode`, check that you dodge the bullets 100% of the time without leaving smoke behind
7. set dodge to 0, enable `Uncanny Dodge Thrusters Mode`, stand inside 3x3 (or larger) smoke/fire, check that you fail to dodge via thrusters
8. set dodge to 0, enable `Uncanny Dodge Thrusters Mode`, enable debug HVAC, stand inside 3x3 (or larger) smoke/fire, check that you can dodge via thrusters

## Describe alternatives you've considered

- go with #4125, that's also perfectly fine solution.
- make thrusters mode separate CBM? (since currently it's a bit awkward as at dodge level 10 thruster mode does nothing)

## Additional Context

![image](<https://github.com/cataclysmbnteam/Cataclysm-BN/assets/54838975/ffae847b-ea9a-43c2-98d2-2d0f2e0a5809> "Immortal Smoke lulz")
